clouddriver-artifacts/src/main/java/com/netflix/spinnaker/clouddriver/artifacts/ivy/IvyArtifactCredentials.java

```diff
@@ -0,0 +1,137 @@
+
+import com.netflix.spinnaker.clouddriver.artifacts.config.ArtifactCredentials;
+import com.netflix.spinnaker.kork.artifacts.model.Artifact;
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.ivy.Ivy;
+import org.apache.ivy.core.module.id.ModuleId;
+import org.apache.ivy.core.module.id.ModuleRevisionId;
+import org.apache.ivy.core.report.ResolveReport;
+import org.apache.ivy.core.resolve.ResolveOptions;
+import org.apache.ivy.util.AbstractMessageLogger;
+import org.apache.ivy.util.Message;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.text.ParseException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.function.Supplier;
+
+@Slf4j
+@Data
+public class IvyArtifactCredentials implements ArtifactCredentials {
+  private final List<String> types = Collections.singletonList("ivy/file");
+  private final IvyArtifactAccount account;
+  private final Supplier<Path> cacheBuilder;
+
+  public IvyArtifactCredentials(IvyArtifactAccount account) {
+    this(account, () -> Paths.get(System.getProperty("java.io.tmpdir"), UUID.randomUUID().toString()));
+  }
+
+  public IvyArtifactCredentials(IvyArtifactAccount account, Supplier<Path> cacheBuilder) {
+    this.cacheBuilder = cacheBuilder;
+    redirectIvyLogsToSlf4j();
+    this.account = account;
+  }
+
+  private static void redirectIvyLogsToSlf4j() {
+    Message.setDefaultLogger(new AbstractMessageLogger() {
+      private final Logger logger = LoggerFactory.getLogger("org.apache.ivy");
+
+      @Override
+      protected void doProgress() {
+      }
+
+      @Override
+      protected void doEndProgress(String msg) {
+        log(msg, Message.MSG_INFO);
+      }
+
+      @Override
+      public void log(String msg, int level) {
+        switch (level) {
+          case Message.MSG_ERR:
+            logger.error(msg);
+            break;
+          case Message.MSG_WARN:
+            logger.warn(msg);
+            break;
+          case Message.MSG_INFO:
+            logger.info(msg);
+            break;
+          case Message.MSG_DEBUG:
+            logger.debug(msg);
+            break;
+          case Message.MSG_VERBOSE:
+            logger.trace(msg);
+          default:
+            // do nothing
+        }
+      }
+
+      @Override
+      public void rawlog(String msg, int level) {
+        log(msg, level);
+      }
+    });
+  }
+
+  public InputStream download(Artifact artifact) {
+    Path cacheDir = cacheBuilder.get();
+    Ivy ivy = account.getSettings().toIvy(cacheDir);
+
+    String[] parts = artifact.getReference().split(":");
+    if (parts.length < 3) {
+      throw new IllegalArgumentException("Ivy artifact reference must have a group, artifact, and version separated by ':'");
+    }
+
+    ModuleRevisionId mrid = new ModuleRevisionId(new ModuleId(parts[0], parts[1]), parts[2]);
```

Comment:
Should we add the (optional) classifier here as well?

Reply:
It's a little trickier than that in Ivy because Maven POMs don't actually list their artifacts. We'll have to build a chain resolver and construct an intermediate Ivy file for classifiers. Something like:

```
<ivy-module version="2.0" xmlns:e="http://ant.apache.org/ivy/extra">
    <info organisation="com.test" module="artifact-with-classifier" revision="1.0"/>

    <configurations defaultconf="master">
        <conf name="master"/>
    </configurations>

    <dependencies>
        <dependency org="com.test" name="artifact" rev="1.0">
            <artifact name="artifact" e:classifier="bundle"/>
        </dependency>
    </dependencies>

</ivy-module>
```

Reply:
Hoping to come back to this next week.